File: apps/web/src/app/protocol/openings.ts
// `GET /api/openings` 의 계약. 서버의 `internal/server/openings.go` 와 짝이다.
//
// **수순이 여기 없다.** 서버가 이름과 한 줄 설명까지만 보낸다 — 상대가 다음에 무엇을 둘지가
// 페이로드에 있으면 devtools 하나로 초반이 통째로 보인다(그쪽 파일의 `openingItem` 주석).

/** 고를 수 있는 진형 하나. */
export interface Opening {
  id: string;
  /** 일본어 이름(四間飛車). 그대로 그린다. */
  name: string;
  /** 한 줄 설명. 초심자가 무엇을 고르는지 알 수 있게 서버가 일본어로 준다. */
  note: string;
  /** 출처 URL. 화면에서 이름에 걸어 둔다 — 인용 규약은 journal §30. */
  source: string;
}

/**
 * 진형 목록을 받는다.
 *
 * **실패하면 빈 목록이다.** 이 자리가 막혀도 「おまかせ」로 대국은 시작할 수 있어야 한다 —
 * 목록은 고르는 즐거움이고 대국이 본체다(엔진·DB가 없어도 판정만 꺼지는 것과 같은 판단).
 */
export async function fetchOpenings(signal: AbortSignal): Promise<Opening[]> {
  try {
    const res = await fetch('/api/openings', { signal });
    if (!res.ok) return [];
    const body = (await res.json()) as { openings?: Opening[] };
    return body.openings ?? [];
  } catch {
    return [];
  }
}

File: apps/server/internal/server/openings.go
package server

import (
	"net/http"

	"github.com/jovid18/show-gi/apps/server/internal/book"
)

// openingItem 은 시작 화면이 그리는 한 줄이다.
//
// 수순을 안 보낸다. 상대가 다음에 무엇을 둘지가 클라이언트에 있으면 devtools 하나로
// 초반이 통째로 보이고, 그건 「최선수를 보여주지 않는다」(01-core.md §1)를 진형 쪽에서
// 뚫는 것이다. 화면에 필요한 것은 이름과 한 줄 설명뿐이다.
type openingItem struct {
	ID   string `json:"id"`
	Name string `json:"name"`
	Note string `json:"note"`
	// Source 는 출처 URL. 화면 아래에 그대로 붙는다 — 인용 규약은 journal §30.
	Source string `json:"source"`
}

func openings(w http.ResponseWriter, _ *http.Request) {
	all := book.All()
	out := make([]openingItem, 0, len(all))
	for _, o := range all {
		out = append(out, openingItem{ID: o.ID, Name: o.Name, Note: o.Note, Source: o.Source})
	}
	writeJSON(w, http.StatusOK, map[string]any{"openings": out})
}
